### PR TITLE
ci: Implement Arduino auto-format into CI flow

### PR DESCRIPTION
**Pull request by alayander**
Implement auto-formatter for Arduino using clang-format

### CODE REVIEW

File: ci/ci_venv.sh
Comment (BrianShTsoi) [inline ci/ci_venv.sh]: nit: We should probably early return here instead of moving on with python?

File: ci/pre-commit
Comment (BrianShTsoi) [inline ci/pre-commit]: I realize that instead of assuming "existing ci_venv" == "all necessary packages are installed", maybe it's better to check if each package exists right before we are using it? That way we can also avoid situations where someone using an old `ci_venv/` that didn't get updated after rebasing?
Not sure how much more effort it requires to change it. Might not be worth the trouble since this is lower priority anyways so it's up to @alayander .

**Review by BrianShTsoi — CHANGES_REQUESTED** [review]
Left some short comments.

Can't help but feel like we could've write it all in one script that takes in some arguments of what files to run CI on (maybe plus a small script for precommit, calling the big script), especially since activating `ci_venv` takes multiple lines and is duplicated 3 times.
But again might not be worth the trouble. I'll leave it up to @alayander .

Otherwise looks good, can probably merge after fixing the nit. Might not even have to wait for @PigeonKing-Kevin considering he has other priorities and PMs are doing the arduino part right now?

File: ci/pre-commit
Comment (alayander) [reply ci/pre-commit]: Resolved this by checking that all packages are installed during activation, and installing otherwise.

**Comment by alayander** [comment]
> Can't help but feel like we could've write it all in one script that takes in some arguments of what files to run CI on (maybe plus a small script for precommit, calling the big script), especially since activating `ci_venv` takes multiple lines and is duplicated 3 times. But again might not be worth the trouble. I'll leave it up to @alayander .

This was my initial thought with a check of environment variables that would be set if the script is called via a git hook vs manually. The functionality is still fundamentally different where it would be a case of branching entirely in an if else block. At that point they are essentially two different scripts, since behaviour branches immediately. Additionally the script would need to be named `pre-commit` for the git hook to be recognized, which wouldn't really make sense when running manually.

**Comment by BrianShTsoi** [comment]
I was more so thinking that we can have one Ci script that takes in arguments of file paths to run CI on (possibly with the default path being `./`).
Then the pre-commit script is just responsible for identifying the staged files and call the CI script with the staged files as args. 
Cuz like the common functionality is "source ci_venv and run CI on some files", and the difference in logic is just what files we want to run CI on

**Comment by alayander** [comment]
> I was more so thinking that we can have one Ci script that takes in arguments of file paths to run CI on (possibly with the default path being `./`). Then the pre-commit script is just responsible for identifying the staged files and call the CI script with the staged files as args. Cuz like the common functionality is "source ci_venv and run CI on some files", and the difference in logic is just what files we want to run CI on

IMO there's more nuance to just running the entire CI suite on the staged files identified. The flow with `pre-commit` has the different stages of CI build up on each other (Ex. If black modifies a file - the script will exit and the commit is already abandoned. Whereas with `ci/run_ci.sh`, the entire suite runs altogether). As well, the `pylint` portion of `pre-commit` is dependent on parsing the output of `pylint`, which would be more tricky to capture unless it gets returned by the CI script or all output is captured, but at that point it doesn't make much sense to me to build out that dependency.

**Comment by BrianShTsoi** [comment]
Ah well, guess that's too much complexity then. 
I'll run it locally again to make sure it's working, then we should be good to merge, unless @PigeonKing-Kevin wants to say something.

**Review by BrianShTsoi — DISMISSED** [review]
LGTM!

**Review by PigeonKing-Kevin — APPROVED** [review]
LGTM!

**Review by BrianShTsoi — APPROVED** [review]
¯\_(ツ)_/¯

**Pull request merged**
